### DISCUSSION TITLE
Defining the expected behavior of wild operations for storages

### DISCUSSION THREAD

**Discussion by p-avital** [eclipse-zenoh/roadmap#26] [Ideas and new features]
A long-standing debate in the team has been the expected behavior of storages on key expressions containing wilds. Here are the two main positions that we have found ourselves gravitating around during these debates.

These debates hinge on the fact that key expressions define sets that may either have size 1 (KE doesn't contain `*`) or be unsized (KE contains at least one `*`).

The pedantic approach would be to treat operations on either type of set as identical, letting the user define a value for an infinite amount of ressources.

The parsimonious approach would be to treat sized and unsized puts as distinct operations: the former defines ressources if they didn't exist yet, otherwise sets their value; the later only sets the value for matching existing ressources

## Wild puts define ressources
This is the "pedantic" approach:
```rust
get("*"); // {}
put("a", 0);
get("*"); // {a: 0}
put("*", 1);
get("*"); // {*: 1} 
put("b", 2);
get("*"); // {*: 1, b: 2} 
get("c"); // {*: 1}
```

## Wild puts update existing ressources
This is the "parsimonious" approach:
```rust
get("*"); // {}
put("a", 0);
get("*"); // {a: 0}
put("*", 1);
get("*"); // {a: 1}
put("b", 2);
get("*"); // {a: 1, b: 2} 
get("c"); // {}
```

### Pedantic
```
+ All operations are treated the same.
+ Storages state is aligned with that of subscribers that may have subscribed
  on specific ressources that are never addressed through sized operations.
- The only way to remove a wild from a storage is to update one of its supersets.
```
### Parsimonious
```
+ On top of pedantic-compatible garbage collection, wild ressources may be deleted
  from storages whenever consistency is acertained.
+ No "surprise ressources" when getting a ressource that's never been addressed
  through sized operations.
+ It is similar to what a user would expect when operating with the * wildcard on Unix-like systems, which will be 
  expanded to the set of actual existing resources created via previous operations. 
- Storages are unaware of the state that some nodes may hold for some ressources.
```

**Comment by Mallets** [comment]
An additional point for the parsimonious case:
```
+ It is similar to what a user would expect when operating with the * wildcard on Unix-like systems, which will be 
  expanded to the set of actual existing resources created via previous operations. 
```
